lib/src/block_parser.dart

```diff
@@ -728,7 +730,137 @@ class ParagraphSyntax extends BlockSyntax {
       parser.advance();
     }
 
-    var contents = parser.document.parseInline(childLines.join('\n'));
-    return new Element('p', contents);
+    var paragraphLines = _extractReflinkDefinitions(parser, childLines);
+    if (paragraphLines == null) {
+      // Paragraph consisted solely of reference link definitions.
+      return new Text('');
+    } else {
+      var contents = new UnparsedContent(paragraphLines.join('\n'));
+      return new Element('p', [contents]);
+    }
+  }
+
+  // Extract reference link definitions from the front of the paragraph, and
+  // return the remaining paragraph lines.
+  List<String> _extractReflinkDefinitions(
+      BlockParser parser, List<String> lines) {
```

Comment:
I think that is similar to my current solution, because it pushes the backtracking complexity into `parseRefLink`.

If `parseRefLink` is given all of the content from some line to the end of the paragraph, it might look like this:

```markdown
[label]: http://url    // already parsed as a link ref def.
[label]: http://url    // [contents] passed to parseRefLink starts here.
"title"
paragraph text.
```

So `parseRefLink` cannot just regexp on the first line and say "oh, `[label]: http://url` looks like a definition! Because it _should_ parse the next line ("title") as well.

However, it might get a few lines in, and then, nope! No title! Like this:

```markdown
[label]:
http://url
"title
with many
words" oops not really a title.
```

Here, `parseRefLink` would have to try like 1 (could be the start of a ref definition), then add 2 (so far so good, but there might be a title), 3 (oops yeah might be a title), 4 (still might), 5 (oops, nope, weird trailing characters; not a title! backtrack to the place earlier with just the label and url)